peps/pep-0748.rst

```diff
@@ -0,0 +1,1573 @@
+a configuration language by most products that use OpenSSL, including Python.
+This format is relatively easy to read, but has a number of downsides: it is a
+string, which makes it remarkably easy to provide bad inputs; it lacks much
+detailed validation, meaning that it is possible to configure OpenSSL in a way
+that doesn't allow it to negotiate any cipher at all; and it allows specifying
+cipher suites in a number of different ways that make it tricky to parse. The
+biggest problem with this format is that there is no formal specification for
+it, meaning that the only way to parse a given string the way OpenSSL would is
+to get OpenSSL to parse it.
+
+OpenSSL's cipher strings can look like this:
+
+.. code-block:: python
+
+    "ECDH+AESGCM:ECDH+CHACHA20:DH+AESGCM:DH+CHACHA20:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!eNULL:!MD5"
+
+
+This string demonstrates some of the complexity of the OpenSSL format. For
+example, it is possible for one entry to specify multiple cipher suites: the
+entry ``ECDH+AESGCM`` means “all ciphers suites that include both elliptic-curve
+Diffie-Hellman key exchange and AES in Galois Counter Mode”. More explicitly,
+that will expand to four cipher suites:
+
+
+.. code-block:: python
+
+    "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256"
+
+
+That makes parsing a complete OpenSSL cipher string extremely tricky. Add to the
+fact that there are other meta-characters, such as “!” (exclude all cipher
+suites that match this criterion, even if they would otherwise be included:
+“!MD5” means that no cipher suites using the MD5 hash algorithm should be
+included), “-” (exclude matching ciphers if they were already included, but
+allow them to be re-added later if they get included again), and “+” (include
+the matching ciphers, but place them at the end of the list), and you get an
+extremely complex format to parse. On top of this complexity it should be noted
+that the actual result depends on the OpenSSL version, as an OpenSSL cipher
+string is valid so long as it contains at least one cipher that OpenSSL
+recognizes.
+
+OpenSSL also uses different names for its ciphers than the names used in the
+relevant specifications. See the manual page for ``ciphers(1)`` for more
+details.
+
+The actual API inside OpenSSL for the cipher string is simple:
+
+.. code-block:: c
+
+    char *cipher_list = <some cipher list>;
+    int rc = SSL_CTX_set_cipher_list(context, cipher_list);
+
+
+This means that any format that is used by this module must be able to be
+converted to an OpenSSL cipher string for use with OpenSSL.
+
+Network Framework
+^^^^^^^^^^^^^^^^^
+
+Network Framework is the macOS (10.15+) system TLS library. This library is
+substantially more restricted than OpenSSL in many ways, as it has a much more
+restricted class of users. One of these substantial restrictions is in
```

Comment:
It seems like it would be worth mentioning that Network Framework precludes the implementation of either `TLSBuffer` or `TLSSocket`, since there is no way to parse bytes into or out of a bytes buffer, nor any way to cause `send` and `recv` to block, as the Network Framework versions of those functions take completion handlers and require a CoreFoundation runloop to already be running.

It's a little misleading to even call Network Framework a "system TLS library", which is why it seems "restricted". Personally I find its architecture very annoying, and badly done in a lot of ways, but in its model it considers TLS to be a the same sort of layer for applications as TCP.  In the same way that you can't get raw sockets and then ask the operating system's ethernet driver to assemble connection data from segments in frames with BSD sockets, you cannot ask the operating system's TLS driver to assemble connection data from TCP streams into TLS records for you.